src/kernel/cpu/binary_reduce_impl.h

```diff
@@ -0,0 +1,228 @@
+          typename BinaryOp, typename Reducer>
+void CallBinaryReduce(const minigun::advance::RuntimeConfig& rtcfg,
+                      const ImmutableGraph* graph,
+                      GData<Idx, DType>* gdata) {
+  typedef cpu::FunctorsTempl<Idx, DType, LeftSelector,
+                        RightSelector, BinaryOp, Reducer>
+          Functors;
+  typedef cpu::BinaryReduce<Idx, DType, Functors> UDF;
+  // csr
+  auto outcsr = graph->GetOutCSR();
+  minigun::Csr<Idx> csr = utils::CreateCsr<Idx>(outcsr->indptr(), outcsr->indices());
+  // If the user-given mapping is none and the target is edge data, we need to
+  // replace the mapping by the edge ids in the csr graph so that the edge
+  // data is correctly read/written.
+  if (LeftSelector::target == binary_op::kEdge && gdata->lhs_mapping == nullptr) {
+    gdata->lhs_mapping = static_cast<Idx*>(outcsr->edge_ids()->data);
+  }
+  if (RightSelector::target == binary_op::kEdge && gdata->rhs_mapping == nullptr) {
+    gdata->rhs_mapping = static_cast<Idx*>(outcsr->edge_ids()->data);
+  }
+  if (OutSelector<Reducer>::Type::target == binary_op::kEdge
+      && gdata->out_mapping == nullptr) {
+    gdata->out_mapping = static_cast<Idx*>(outcsr->edge_ids()->data);
+  }
+  // TODO(minjie): allocator
+  minigun::advance::Advance<XPU, Idx, cpu::AdvanceConfig, GData<Idx, DType>, UDF>(
+        rtcfg, csr, gdata, minigun::IntArray1D<Idx>());
+}
+
+template <int XPU, int NDim, typename Idx, typename DType,
+          typename LeftSelector, typename RightSelector,
+          typename BinaryOp, typename Reducer>
+void CallBinaryReduceBcast(
+  const minigun::advance::RuntimeConfig& rtcfg,
+  const ImmutableGraph* graph,
+  BcastGData<NDim, Idx, DType>* gdata) {
+  typedef cpu::FunctorsTempl<Idx, DType, LeftSelector,
+                        RightSelector, BinaryOp, Reducer>
+          Functors;
+  typedef cpu::BinaryReduceBcast<NDim, Idx, DType, Functors> UDF;
+  // csr
+  auto outcsr = graph->GetOutCSR();
+  minigun::Csr<Idx> csr = utils::CreateCsr<Idx>(outcsr->indptr(), outcsr->indices());
+  // If the user-given mapping is none and the target is edge data, we need to
+  // replace the mapping by the edge ids in the csr graph so that the edge
+  // data is correctly read/written.
+  if (LeftSelector::target == binary_op::kEdge && gdata->lhs_mapping == nullptr) {
+    gdata->lhs_mapping = static_cast<Idx*>(outcsr->edge_ids()->data);
+  }
+  if (RightSelector::target == binary_op::kEdge && gdata->rhs_mapping == nullptr) {
+    gdata->rhs_mapping = static_cast<Idx*>(outcsr->edge_ids()->data);
+  }
+  if (OutSelector<Reducer>::Type::target == binary_op::kEdge
+      && gdata->out_mapping == nullptr) {
+    gdata->out_mapping = static_cast<Idx*>(outcsr->edge_ids()->data);
+  }
+  // TODO(minjie): allocator
+  minigun::advance::Advance<XPU, Idx, cpu::AdvanceConfig,
+    BcastGData<NDim, Idx, DType>, UDF>(
+        rtcfg, csr, gdata, minigun::IntArray1D<Idx>());
```

Comment:
Question about minigun: is output frontier being used in DGL in any way?  I noticed that a chunk of code in minigun fills in the output frontier buffer which seems to be discarded afterwards.

Reply:
No, it is not. Currently, for subgraph update, we generate explicit subgraph and perform full advance on it. The partial frontier could be useful to speedup `pull` so no explicit subgraph is needed.